### PR TITLE
Fix internal state machine of action node

### PR DESCRIPTION
**Pull request by Oberacda**
The implicit state machine in the action node was not implemented correctly.
This PR implements the state machine explicitly and fixes the invalid transitions in the previous one.

### CODE REVIEW

File: ros_bt_py/src/ros_bt_py/ros_nodes/action.py [hunk 1]
```diff
@@ -164,106 +195,148 @@ def _do_setup(self):
...
         if seconds_running > self.options["timeout_seconds"]:
             self.logwarn(f"Cancelling goal after {seconds_running:f} seconds!")
             self._running_goal_future.cancel()
             self._running_goal_future = None
-            status = self._do_tick_cancel_running_goal()
-            if status != NodeMsg.SUCCEED:
-                return status
+            self._internal_state = ActionStates.REQUEST_GOAL_CANCELLATION
+            return NodeMsg.SUCCEED
+
```
Comment (nspielbau) [inline ros_bt_py/src/ros_bt_py/ros_nodes/action.py:248]: We assume this cancellation does work all the time?

Comment (nspielbau) [reply ros_bt_py/src/ros_bt_py/ros_nodes/action.py:248]: Also shouldn't a timeout lead to a NodeMsg.FAILED?

File: ros_bt_py/src/ros_bt_py/ros_nodes/action.py [hunk 2]
```diff
@@ -164,106 +195,148 @@ def _do_setup(self):
...
+            self.logwarn(
+                "Goal cancellation was requested, but there is no handle to the running goal!"
+            )
+            self._internal_state = ActionStates.FINISHED
             return NodeMsg.BROKEN
+
         self._cancel_goal_future = self._running_goal_handle.cancel_goal_async()
-        return NodeMsg.RUNNING
+        self._internal_state = ActionStates.WAITING_FOR_GOAL_CANCELLATION
+        return NodeMsg.SUCCEED
```
Comment (nspielbau) [inline ros_bt_py/src/ros_bt_py/ros_nodes/action.py:261]: Same as above

File: ros_bt_py/src/ros_bt_py/ros_nodes/action.py [hunk 3]
```diff
@@ -164,106 +195,148 @@ def _do_setup(self):
...
             self._active_goal = None
-            return NodeMsg.FAIL
+
+            self._internal_state = ActionStates.FINISHED
+            return NodeMsg.FAILURE
         return NodeMsg.RUNNING
 
     def _do_tick_send_new_goal(self) -> str:
-        if self._running_goal_handle is not None or self._new_goal_future is not None:
-            self.logfatal("Running goal handle handle or new goal future is not none! Cannot continue!")
```
Comment (nspielbau) [inline ros_bt_py/src/ros_bt_py/ros_nodes/action.py:230]: handle handle?

File: ros_bt_py/src/ros_bt_py/ros_nodes/action.py [hunk 4]
```diff
@@ -164,106 +195,148 @@ def _do_setup(self):
...
+        self._new_goal_request_future = self._ac.send_goal_async(
             goal=self._input_goal, feedback_callback=self._feedback_cb
         )
-        if self._input_goal is not None:
-            self._active_goal = self._input_goal
+
+        self._active_goal = self._input_goal
+        self._internal_state = ActionStates.WAITING_FOR_GOAL_ACCEPTANCE
+
         return NodeMsg.SUCCEED
```
Comment (nspielbau) [inline ros_bt_py/src/ros_bt_py/ros_nodes/action.py:301]: Why do we succeed here? The Action is not done isn't it?

File: ros_bt_py/src/ros_bt_py/ros_nodes/action.py (hunk 1 above)
Comment (Oberacda) [reply ros_bt_py/src/ros_bt_py/ros_nodes/action.py:248]: You are right, in fact we should cancel this goal asynchronously. In regards to the second point we also should return RUNNING to get into the REQUEST_GOAL_CANCELLATION state with the next tick.

File: ros_bt_py/src/ros_bt_py/ros_nodes/action.py (hunk 2 above)
Comment (Oberacda) [reply ros_bt_py/src/ros_bt_py/ros_nodes/action.py:261]: In fact this is done correctly here.
The goal is cancelled asynchronously, thus we need to check the result later (in the WAITING_FOR_GOAL_CANCELLATION state).
Secondly the SUCCESS status indicates that the cancellation request was made successfully.
This is not returned for the tick but used to determine if we need to evaluate another internal state or exit the entire tick prematurely:
https://github.com/fzi-forschungszentrum-informatik/ros2_ros_bt_py/blob/c30152f33d163fb66b7d6e2114cbe52f3578b8cc/ros_bt_py/src/ros_bt_py/ros_nodes/action.py#L377

File: ros_bt_py/src/ros_bt_py/ros_nodes/action.py (hunk 4 above)
Comment (Oberacda) [reply ros_bt_py/src/ros_bt_py/ros_nodes/action.py:301]: As above, for some of the _do_tick methods, the return status does not indicate the entire tick succeeded, but this internal state succeeded. In this case once the goal was successfully submitted this substate succeeds.

**Review by nspielbau — APPROVED** [review]
LGTM

**Pull request merged**